packages/SwingSet/src/kernel/vat-loader/async-helper.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * @template T
+ * @typedef {T extends PromiseLike<any> ? never : T} SyncOnly
+ */
+
+/**
+ * @template {boolean} AllowAsync
+ * @template T
+ * @typedef {true extends AllowAsync ? (Promise<T> | T) : T} MaybePromiseResult
+ */
+
+/**
+ * WARNING: Do not use, this is releasing Zalgo
+ *
+ * @template T
+ * @template R
+ * @param {T} v
+ * @param {(v: Awaited<T>) => R} handle
+ * @returns {T extends PromiseLike<any> ? Promise<Awaited<R>> : R}
+ */
+export function maybeAsync(v, handle) {
+  if (typeof v === 'object' && v !== null && 'then' in v) {
```

Comment:
Consider matching the rules for testing whether it is thenable. I'm not sure the following is exactly right, but it is closer.

```suggestion
  if (!isPrimitive(v) && typeof v.then === 'function') {
```

IOW the presence of `then` is only disqualifying if its value is a function.

v is a candidate whether it is an object or a function.